inc/class-previews.php

```diff
@@ -0,0 +1,153 @@
+ * Irving.
+ */
+class Previews {
+
+	/**
+	 * Class instance.
+	 *
+	 * @var null|self
+	 */
+	protected static $instance;
+
+	/**
+	 * Get class instance
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+		if ( ! isset( static::$instance ) ) {
+			static::$instance = new static();
+			static::$instance->setup();
+		}
+		return static::$instance;
+	}
+
+	/**
+	 * Setup the singleton. Validate JWT is installed, and setup hooks.
+	 */
+	public function setup() {
+
+		// Allow-list some query vars.
+		add_filter( 'query_vars', [ $this, 'modify_query_vars' ] );
+
+		// Filter to easily enable public previews.
+		if ( apply_filters( 'wp_irving_enable_public_previews', '__return_false' ) ) {
+			$this->enable_public_previews();
+		}
+
+		// Validate that JWT exists and is enabled.
+		if ( ! defined( 'JWT_AUTH_VERSION' ) ) {
+			return;
+		}
+
+		// Hook into Irving's query string to WP_Query process to modify some
+		// logic for previews.
+		add_action( 'wp_irving_components_wp_query', [ $this, 'modify_wp_query_for_previews' ], 10, 4 );
+	}
+
+	/**
+	 * Allow-list some query vars for previews.
+	 *
+	 * @param  array $vars Allow-listed query vars.
+	 * @return array
+	 */
+	public function modify_query_vars( array $vars ): array {
+		$vars[] = 'preview_id';
+		$vars[] = 'preview_nonce';
+		return $vars;
+	}
+
+	/**
+	 * Enable public previews.
+	 */
+	public function enable_public_previews() {
+
+		// Disable revision preview nonces.
+		remove_action( 'init', '_show_post_preview' );
+
+		// Hook into the component API request and change the draft preview
+		// to be `publish`, mimicking a public post.
+		add_filter(
+			'wp_irving_components_request',
+			function ( $request ) {
+				if (
+					true === (bool) $request->get_param( 'preview' )
+					|| 0 !== absint( $request->get_param( 'p' ) )
+				) {
+					add_filter(
+						'posts_results',
+						function ( $posts ) {
+
+							if ( empty( $posts ) ) {
+								return $posts;
+							}
+
+							$posts[0]->post_status = 'publish';
+							return $posts;
+						}
+					);
+				}
+
+			}
+		);
+	}
+
+	/**
+	 * Modify the \WP_Query object that Irving creates to swap out the published post, for the correct revision.
+	 *
+	 * @param \WP_Query $wp_query      WP_Query object corresponding to this
+	 *                                 request.
+	 * @param string    $path          Request path.
+	 * @param string    $custom_params Custom params.
+	 * @param string    $params        Request params.
+	 * @return \WP_Qury $wp_query WP_Query object for the request.
+	 */
+	public function modify_wp_query_for_previews( $wp_query, $path, $custom_params, $params ) {
+
+		// Disable revision preview nonces so we can inject our own logic.
+		// Without this, wp_die() will execute, which means our auth strategy
+		// fails. There isn't anyway to currently filter this behavior, so
+		// we're stuck with this.
+		remove_action( 'init', '_show_post_preview' );
```

Comment:
This is happening too late, I've already hit the `wp_die()` message of `Sorry, you are not allowed to preview drafts` before it gets here to remove that action. I'd move this line to the setup function since it's used in both cases (authed previews and public previews).

Reply:
Also a good call. Updated.